File: backend/src/main/java/org/luxons/sevenwonders/game/resources/BestPriceCalculator.java
package org.luxons.sevenwonders.game.resources;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.boards.Board;

public class BestPriceCalculator {

    private final Resources resources;

    private final List<ResourcePool> pools;

    private BestPriceCalculator(Resources resources, Table table, int playerIndex) {
        this.resources = resources;
        this.pools = createResourcePools(table, playerIndex);
    }

    private static List<ResourcePool> createResourcePools(Table table, int playerIndex) {
        Provider[] providers = Provider.values();
        List<ResourcePool> pools = new ArrayList<>(providers.length + 1);

        Board board = table.getBoard(playerIndex);
        TradingRules rules = board.getTradingRules();
        pools.add(new ResourcePool(board.getProduction(), null, rules));

        for (Provider provider : providers) {
            Board providerBoard = table.getBoard(playerIndex, provider.getBoardPosition());
            ResourcePool pool = new ResourcePool(providerBoard.getPublicProduction(), provider, rules);
            pools.add(pool);
        }
        return pools;
    }

    public static int bestPrice(Resources resources, Table table, int playerIndex) {
        Board board = table.getBoard(playerIndex);
        Resources leftToPay = resources.minus(board.getProduction().getFixedResources());
        return new BestPriceCalculator(leftToPay, table, playerIndex).bestPrice();
    }

    private int bestPrice() {
        if (resources.isEmpty()) {
            return 0;
        }
        int currentMinPrice = Integer.MAX_VALUE;
        for (ResourceType type : ResourceType.values()) {
            if (resources.getQuantity(type) > 0) {
                int minPriceUsingOwnResource = bestPriceWithout(type);
                currentMinPrice = Math.min(currentMinPrice, minPriceUsingOwnResource);
            }
        }
        return currentMinPrice;
    }

    private int bestPriceWithout(ResourceType type) {
        resources.remove(type, 1);
        int currentMinPrice = Integer.MAX_VALUE;
        for (ResourcePool pool : pools) {
            int resCostInPool = pool.getCost(type);
            for (Set<ResourceType> choice : pool.getChoices()) {
                if (choice.contains(type)) {
                    Set<ResourceType> temp = EnumSet.copyOf(choice);
                    choice.clear();
                    currentMinPrice = Math.min(currentMinPrice, bestPrice() + resCostInPool);
                    choice.addAll(temp);
                }
            }
        }
        resources.add(type, 1);
        return currentMinPrice;
    }

    private static class ResourcePool {

        private final Set<Set<ResourceType>> choices;

        private final Provider provider;

        private final TradingRules rules;

        private ResourcePool(Production production, Provider provider, TradingRules rules) {
            this.choices = production.asChoices();
            this.provider = provider;
            this.rules = rules;
        }

        Set<Set<ResourceType>> getChoices() {
            return choices;
        }

        int getCost(ResourceType type) {
            if (provider == null) {
                return 0;
            }
            return rules.getCost(type, provider);
        }
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/game/resources/BestPriceCalculator.kt
package org.luxons.sevenwonders.game.resources

import org.luxons.sevenwonders.game.Player
import org.luxons.sevenwonders.game.api.resources.Provider
import org.luxons.sevenwonders.game.api.resources.ResourceTransactions
import org.luxons.sevenwonders.game.api.resources.ResourceType
import java.util.EnumSet

internal fun bestSolution(resources: Resources, player: Player): TransactionPlan =
    BestPriceCalculator(resources, player).computeBestSolution()

data class TransactionPlan(val price: Int, val possibleTransactions: Set<ResourceTransactions>)

private class ResourcePool(
    val provider: Provider?,
    private val rules: TradingRules,
    choices: Set<Set<ResourceType>>
) {
    val choices: Set<MutableSet<ResourceType>> = choices.map { it.toMutableSet() }.toSet()

    fun getCost(type: ResourceType): Int = if (provider == null) 0 else rules.getCost(type, provider)
}

private class BestPriceCalculator(resourcesToPay: Resources, player: Player) {

    private val pools: List<ResourcePool>
    private val resourcesLeftToPay: MutableResources
    private val boughtResources: MutableMap<Provider, MutableResources> = HashMap()
    private var pricePaid: Int = 0

    private var bestSolutions: MutableSet<ResourceTransactions> = mutableSetOf()
    private var bestPrice: Int = Integer.MAX_VALUE

    init {
        val board = player.board
        this.resourcesLeftToPay = resourcesToPay.minus(board.production.getFixedResources()).toMutableResources()
        this.pools = createResourcePools(player)
    }

    private fun createResourcePools(player: Player): List<ResourcePool> {
        // we only take alternative resources here, because fixed resources were already removed for optimization
        val ownBoardChoices = player.board.production.getAlternativeResources()
        val ownPool = ResourcePool(null, player.board.tradingRules, ownBoardChoices)
        val providerPools = Provider.values().map { it.toResourcePoolFor(player) }

        return providerPools + ownPool
    }

    private fun Provider.toResourcePoolFor(player: Player): ResourcePool {
        val providerBoard = player.getBoard(boardPosition)
        val choices = providerBoard.publicProduction.asChoices()
        return ResourcePool(this, player.board.tradingRules, choices)
    }

    fun computeBestSolution(): TransactionPlan {
        computePossibilities()
        return TransactionPlan(bestPrice, bestSolutions)
    }

    private fun computePossibilities() {
        if (resourcesLeftToPay.isEmpty()) {
            updateBestSolutionIfNeeded()
            return
        }
        for (type in ResourceType.values()) {
            if (resourcesLeftToPay[type] > 0) {
                for (pool in pools) {
                    if (pool.provider == null) {
                        computeSelfPossibilities(type, pool)
                    } else {
                        computeNeighbourPossibilities(pool, type, pool.provider)
                    }
                }
            }
        }
    }

    private fun computeSelfPossibilities(type: ResourceType, pool: ResourcePool) {
        resourcesLeftToPay.remove(type, 1)
        computePossibilitiesWhenUsing(type, pool)
        resourcesLeftToPay.add(type, 1)
    }

    private fun computeNeighbourPossibilities(pool: ResourcePool, type: ResourceType, provider: Provider) {
        val cost = pool.getCost(type)
        resourcesLeftToPay.remove(type, 1)
        buyOne(provider, type, cost)
        computePossibilitiesWhenUsing(type, pool)
        unbuyOne(provider, type, cost)
        resourcesLeftToPay.add(type, 1)
    }

    fun buyOne(provider: Provider, type: ResourceType, cost: Int) {
        boughtResources.getOrPut(provider) { MutableResources() }.add(type, 1)
        pricePaid += cost
    }

    fun unbuyOne(provider: Provider, type: ResourceType, cost: Int) {
        pricePaid -= cost
        boughtResources.get(provider)!!.remove(type, 1)
    }

    private fun computePossibilitiesWhenUsing(type: ResourceType, pool: ResourcePool) {
        for (choice in pool.choices) {
            if (choice.contains(type)) {
                val temp = EnumSet.copyOf(choice)
                choice.clear()
                computePossibilities()
                choice.addAll(temp)
            }
        }
    }

    private fun updateBestSolutionIfNeeded() {
        if (pricePaid > bestPrice) return

        if (pricePaid < bestPrice) {
            bestPrice = pricePaid
            bestSolutions.clear()
        }
        // avoid mutating the resources from the transactions
        val transactionSet = boughtResources.mapValues { (_, res) -> res.copy() }.toTransactions()
        bestSolutions.add(transactionSet)
    }
}
